lib/protocol/encryption.go

```diff
@@ -0,0 +1,521 @@
+		}
+		data = nd
+	}
+	fileKey := FileKey(realName, folderKey)
+	enc := encryptBytes(data, fileKey)
+	resp.Close()
+	return rawResponse{enc}, nil
+}
+
+func (e encryptedModel) DownloadProgress(deviceID DeviceID, folder string, updates []FileDownloadProgressUpdate) error {
+	if _, ok := e.folderKeys[folder]; !ok {
+		return e.Model.DownloadProgress(deviceID, folder, updates)
+	}
+
+	// The updates contain nonsense names and sizes, so we ignore them.
+	return nil
+}
+
+func (e encryptedModel) ClusterConfig(deviceID DeviceID, config ClusterConfig) error {
+	// TODO: Filter/clean the incoming ClusterConfig?
+	return e.Model.ClusterConfig(deviceID, config)
+}
+
+// The encryptedConnection sits between the model and the encrypted device. It
+// encrypts outgoing metadata and decrypts incoming responses.
+type encryptedConnection struct {
+	Connection
+	folderKeys map[string]*[keySize]byte // folder ID -> key
+}
+
+func (e encryptedConnection) Index(ctx context.Context, folder string, files []FileInfo) error {
+	if folderKey, ok := e.folderKeys[folder]; ok {
+		encryptFileInfos(files, folderKey)
+	}
+	return e.Connection.Index(ctx, folder, files)
+}
+
+func (e encryptedConnection) IndexUpdate(ctx context.Context, folder string, files []FileInfo) error {
+	if folderKey, ok := e.folderKeys[folder]; ok {
+		encryptFileInfos(files, folderKey)
+	}
+	return e.Connection.IndexUpdate(ctx, folder, files)
+}
+
+func (e encryptedConnection) Request(ctx context.Context, folder string, name string, blockNo int, offset int64, size int, hash []byte, weakHash uint32, fromTemporary bool) ([]byte, error) {
+	folderKey, ok := e.folderKeys[folder]
+	if !ok {
+		return e.Connection.Request(ctx, folder, name, blockNo, offset, size, hash, weakHash, fromTemporary)
+	}
+
+	// Encrypt / adjust the request parameters.
+
+	origSize := size
+	if size < minPaddedSize {
+		// Make a request for minPaddedSize data instead of the smaller
+		// block. We'll chop of the extra data later.
+		size = minPaddedSize
+	}
+	encName := encryptName(name, folderKey)
+	encOffset := offset + int64(blockNo*blockOverhead)
+	encSize := size + blockOverhead
+
+	// Perform that request, getting back and encrypted block.
+
+	bs, err := e.Connection.Request(ctx, folder, encName, blockNo, encOffset, encSize, nil, 0, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return the decrypted block (or an error if it fails decryption)
+
+	fileKey := FileKey(name, folderKey)
+	bs, err = DecryptBytes(bs, fileKey)
+	if err != nil {
+		return nil, err
+	}
+	return bs[:origSize], nil
+}
+
+func (e encryptedConnection) DownloadProgress(ctx context.Context, folder string, updates []FileDownloadProgressUpdate) {
+	if _, ok := e.folderKeys[folder]; !ok {
+		e.Connection.DownloadProgress(ctx, folder, updates)
+	}
+
+	// No need to send these
+}
+
+func (e encryptedConnection) ClusterConfig(config ClusterConfig) {
+	// TODO: Filter/clean the outgoing ClusterConfig?
+	e.Connection.ClusterConfig(config)
```

Comment:
Is the TODO still relevant?

Reply:
My thought here was that we might want to be more restrictive in a clusterconfig towards an untrusted device, perhaps not mentioning other devices at all like we do normally.